services/nexus/nexus.service.js

```diff
@@ -157,7 +182,23 @@ module.exports = class Nexus extends BaseJsonService {
     })
   }
 
-  async fetch({ server, repo, groupId, artifactId, queryOpt }) {
+  async fetch({ server, repo, groupId, artifactId, queryOpt, nexus3 }) {
+    if (typeof nexus3 !== 'undefined') {
+      return this.fetch3({ server, repo, groupId, artifactId, queryOpt })
+    }
+    // Most servers still use Nexus 2. Fall back to Nexus 3 if the hitting a
+    // Nexus 2 endpoint returns a Bad Request (=> InvalidResponse).
+    try {
+      return await this.fetch2({ server, repo, groupId, artifactId, queryOpt })
```

Comment:
In the invalid response on the v2 endpoint scenario, what would you think about throw a custom error to the users to let them know they need the `nexus3` flag instead of trying again with v3? Maybe something like this down below (probably with a better message 😄) :

```js
      if (e instanceof InvalidResponse) {
        throw new InvalidResponse({ prettyMessage: 'v2 endpoint not supported' })
      }
```

The request for data seems to take longer (relative to the other platform/data providers we integrate with), especially with https://oss.sonatype.org (which I believe is the main/most popular Nexus instance) and I worry we could see some latency/timing issues.

Reply:
Well, I was trying to make sure that badges would keep on working without user intervention when a given Nexus instance gets migrated from 2 to 3 (see the discussions in #4410).

If/when https://oss.sonatype.org is migrated, throwing a custom error would effectively break _a lot_ of badges, whereas the approach in this PR will maintain functionality automatically. When the migration happens, we can swap over the code to prefer Nexus 3 endpoints, ignore the optional `nexus3` parameter and instead add an optional `nexus2` parameter. Like that we ensure that all users keep on having working badges and that a majority of users (using https://oss.sonatype.org) have optimal performance. 😉

Reply:
I'd still prefer that we fail loudly on an error and defer to users to be explicit. This was the tact taken with the Sonar badges that also have two very different API sets depending on the version of the instance. 

I agree that doing so means when a target Nexus instance is upgraded that the badges would break, including whenever the main Sonatype instance is upgraded. However, as you mentioned, we'll know within 12 hours of that happening (service tests) and can act accordingly. 

I don't care that much, so I won't be a blocker if others are good with this, but that's my two cents

Reply:
Hmmm, I didn't know about the Sonar case, support for the new API was added back in November 2017 when I wasn't yet a maintainer. Both approaches are implemented in our current code base: the F-Droid badge follows a try/fallback logic similar to this PR (the underlying document could either be an old .txt or a newer .yml format). See @paulmelnikow  and my discussions here: https://github.com/badges/shields/pull/2377#issuecomment-441385916

Let's see what other people think about this Nexus case and let's try to be more consistent in the future!

Reply:
> Let's see what other people think about this Nexus case and let's try to be more consistent in the future! 👍

Definitely! Dealing with upstream API changes is something I've been wondering about lately anyway so glad to have discussion on the topic (though no need to necessarily block this as what you have may be the best way to minimize future impacts on these badges). 

More generally, I completely understand why we need be careful to avoid breaking existing badges (especially since we don't have any versioning ourselves), but it feels like a really tall order for Shields to have to maintain backwards _and_ future compatibility with all the platforms we integrate with on behalf of users, especially when those external systems can change their APIs at will.